src/centrality/truss.cpp

```diff
@@ -0,0 +1,254 @@
+  igraph_truss_i_unpack(&triangles, &unpacked_triangles);
+  igraph_vector_int_destroy(&triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Get the edge ids of the unpacked triangles. Note: a given eid can occur
+  // multiple times in this list if it is in multiple triangles.
+  IGRAPH_CHECK(igraph_vector_resize(&eid, num_triangles));
+  IGRAPH_CHECK(igraph_get_eids(graph, &eid, &unpacked_triangles, 0, 0, 1));
+  igraph_vector_destroy(&unpacked_triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the support of the edges.
+  igraph_truss_i_compute_support(&eid, &support);
+  igraph_vector_destroy(&eid);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the truss of the edges.
+  IGRAPH_CHECK(igraph_vector_int_resize(truss, igraph_ecount(graph)));
+  IGRAPH_CHECK(igraph_i_trussness(graph, &support, truss));
+  igraph_vector_int_destroy(&support);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  return IGRAPH_SUCCESS;
+}
+
+// Unpack the triangles as a vector of vertices to be a vector of edges.
+// So, instead of the triangle specified as vertices [1, 2, 3], return the
+// edges as [1, 2, 1, 3, 2, 3] so that the support can be computed.
+static void igraph_truss_i_unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri) {
+  long int i, j;
+  long int m = igraph_vector_int_size(tri);
+  for (i = 0, j = 0; i < m; i += 3, j += 6){
+    VECTOR(*unpacked_tri)[j]   = VECTOR(*unpacked_tri)[j+2] = VECTOR(*tri)[i];
+    VECTOR(*unpacked_tri)[j+1] = VECTOR(*unpacked_tri)[j+4] = VECTOR(*tri)[i+1];
+    VECTOR(*unpacked_tri)[j+3] = VECTOR(*unpacked_tri)[j+5] = VECTOR(*tri)[i+2];
+  }
+}
+
+// Compute the edge support, i.e. number of triangles each edge occurs in.
+// Time complexity: O(m), where m is the number of edges listed in eid.
+static void igraph_truss_i_compute_support(const igraph_vector_t *eid, igraph_vector_int_t *support) {
+  long int m = igraph_vector_size(eid);
+  for (long int i = 0; i < m; ++i){
+    VECTOR(*support)[(long int) VECTOR(*eid)[i]] += 1;
+  }
+}
+
+
+/* internal function doing the computations once the support is defined */
+static int igraph_i_trussness(const igraph_t *graph, igraph_vector_int_t *support,
+  igraph_vector_int_t *truss){
+
+  igraph_integer_t fromVertex, toVertex, e1, e2;
+  igraph_vector_bool_t completed;
+  igraph_adjlist_t adjlist;
+  igraph_vector_int_t *fromNeighbors, *toNeighbors, *q1, *q2;
+  igraph_vector_int_t commonNeighbors;
+  bool e1_complete, e2_complete;
+  int j, n, level, newLevel;
+  long int i, seed, ncommon;
+  long int nedges = igraph_vector_int_size(support);
+
+  // Get max possible value = max entry in support.
+  int max = igraph_vector_int_max(support);
+
+  // The vector of levels. Each level of the vector is a set of edges initially
+  // at that level of support, where support is # of triangles the edge is in.
+  vector< unordered_set<long int> > vec(max + 1);
+
+  // Initialize completed edges.
+  IGRAPH_VECTOR_BOOL_INIT_FINALLY(&completed, igraph_ecount(graph));
```

Comment:
I was wondering whether we should use an `std::vector<bool>` here if we are in C++-land already. `igraph_vector_bool_t` is not optimized at all; it uses one byte per item, while I would expect that `std::vector<bool>` is space-optimized in any reasonably implemented standard C++ library.

@szhorvat Any opinions here?